### DISCUSSION TITLE
Testing Interceptors in Isolation

### DISCUSSION THREAD

**Discussion by bhollis** [connectrpc/connect-go#708] [General]
Hi there, we're switching from the Google gRPC libraries to Connect, and so far it's going well. However, we've hit a bit of a snag while trying to migrate over all of our custom interceptors. As far as I can tell, there's no way for us to construct a `request.AnyRequest` ourselves with a populated `spec` field, which makes it difficult for us to test an interceptor in isolation. Our interceptor is only reading `req.Spec().Procedure` and with the Google libraries, it was easy for us to construct a `grpc.UnaryServerInfo` struct to pass in with all kinds of different method names. With Connect, we can't quite figure out any way to do this. As the docs on `AnyRequest` say:

> ```
> // To preserve our ability to add methods to this interface without breaking
> // backward compatibility, only types defined in this package can implement
> // AnyRequest.
> ```

As an example, our original tests looked something like this, where `authz.NewInterceptor` returned a `grpc.UnaryServerInterceptor`:

```go
interceptor := authz.NewInterceptor()
info := &grpc.UnaryServerInfo{
  FullMethod: test.method,
}
_, err := interceptor(
  ctx,
  nil,
  info,
  func(_ context.Context, _ any) (any, error) { return nil, nil },
)
assert.ErrorIs(t, err, test.expectedErr)
```

**Comment by jhump** [comment]
@bhollis, you can provide your own implementation of `AnyRequest`, but it _must_ embed a `*connect.Request`. So you could do something like so:
```go
type testRequest[T any] struct {
	*connect.Request[T]
	method string
}

func (r testRequest[_]) Spec() connect.Spec {
	return connect.Spec{
		Procedure: r.method,
	}
}

func newRequest[T any](msg *T, method string) connect.AnyRequest {
	return testRequest[T]{Request: connect.NewRequest(msg), method: method}
}
```

Note that such a test request will cause problems if passed to the actual Connect framework, so if your interceptor implementation actually needs to _override_ any of these request properties (and pass the modified request to the next interceptor or handler in the chain), that is not currently possible.

But if you are just testing your interceptor in isolation, this technique should be fine.

**Reply by bhollis** [reply]
Awesome, thank you.
